raiden-ts/src/transfers/epics/withdraw.ts

```diff
@@ -1,78 +1,370 @@
+        // if withdrawMessage.success (WithdrawConfirmation) go through, don't fail
+        takeUntil(
+          action$.pipe(filter(isResponseOf<typeof withdrawMessage>(withdrawMessage, action.meta))),
+        ),
+      ),
+    ),
+  );
+
+/**
+ * Upon WithdrawConfirmation, send the on-chain withdraw request
+ *
+ * @param action$ - Observable of newBlock actions
+ * @param state$ - Observable of RaidenStates
+ * @param deps - Epics dependencies
+ * @param deps.address - Our address
+ * @param deps.signer - Signer instance
+ * @param deps.main - Main signer and address (if present)
+ * @param deps.log - Logger instance
+ * @param deps.getTokenNetworkContract - TokenNetwork contract getter
+ * @param deps.latest$ - Latest observable
+ * @returns Observable of withdrawExpired actions
+ */
+export const withdrawSendTxEpic = (
+  action$: Observable<RaidenAction>,
+  {}: Observable<RaidenState>,
+  { address, signer, main, log, getTokenNetworkContract, latest$ }: RaidenEpicDeps,
+): Observable<withdraw.success | withdraw.failure> =>
+  action$.pipe(
+    filter(withdrawMessage.success.is),
+    filter((action) => action.meta.direction === Direction.SENT),
+    concatMap((action) =>
+      latest$.pipe(
+        first(),
+        mergeMap(({ state, config }) => {
+          const { subkey: configSubkey, revealTimeout } = config;
+          // don't send on-chain tx if we're 'revealTimeout' blocks from expiration
+          // this is our confidence threshold when we can get a tx inside timeout
+          assert(
+            action.meta.expiration >= state.blockNumber + revealTimeout,
+            ErrorCodes.WITHDRAW_EXPIRES_SOON,
+          );
+          const channel = state.channels[channelKey(action.meta)];
+          assert(channel?.state === ChannelState.open, 'channel not open');
+          assert(action.meta.totalWithdraw.gt(channel.own.withdraw), 'withdraw already performed');
+          const req = channel.own.pendingWithdraws.find(
+            (req) =>
+              req.type === MessageType.WITHDRAW_REQUEST &&
+              req.total_withdraw.eq(action.payload.message.total_withdraw) &&
+              req.expiration.eq(action.payload.message.expiration),
+          );
+          assert(req, 'no matching WithdrawRequest found');
+          const { tokenNetwork } = action.meta;
+          const { signer: onchainSigner } = chooseOnchainAccount(
+            { signer, address, main },
+            configSubkey,
+          );
+          const tokenNetworkContract = getContractWithSigner(
+            getTokenNetworkContract(tokenNetwork),
+            onchainSigner,
+          );
+
+          return tokenNetworkContract.functions.setTotalWithdraw(
+            channel.id,
+            address,
+            action.meta.totalWithdraw,
+            action.meta.expiration,
+            req.signature,
+            action.payload.message.signature,
+          );
```

Comment:
This looks like a common issue for many of on-chain transactions. Depending on how often it would make sense to abstract it. Instead have a wrapper which you tell to do a a `setTotalWithdraw` transaction for a contract with address xyz. There could be then even another layer inside to determine the signer. This epic should don't care about such low-level details. And since this is probably part of more than just one epic...

Reply:
Calling this specific tx function in a very specific way is basically the whole job of this epic, that's why it's here. Signer and function are the least of our problems, usually there're a lot of things to get right before getting a contract call like that, specially about parameters, and so far it made sense to be explicit on the epic itself. But we can look into having a service for that, thank you for raising this question.